src/crypto/tls/handshake_server.go

```diff
@@ -136,6 +136,20 @@ func (c *Conn) readClientHello() (*clientHelloMsg, error) {
 		return nil, unexpectedMessageError(clientHello, msg)
 	}
 
+	// TODO(cjpatton): DON'T MERGE: Resolve the following comment from
+	// @Lekensteyn:
+	//
+	// Because this is called before GetConfigForClient() (see below), ECH
+	// cannot selectively be enabled/disabled or controlled based on the outer
+	// CH parameters such as SNI or things like the destination IP address. So
+	// we cannot limit ECH to a single server name only, it would be
+	// automatically enabled for all requests. Just checking that this is
+	// intentional?
+	clientHello, err = c.echAcceptOrBypass(clientHello)
+	if err != nil {
+		return nil, fmt.Errorf("tls: %s", err) // Alert sent.
+	}
```

Comment:
Also, if ECH is rejected, we need to ignore (and not call `checkForResumption()`), right? I think that code path is still hit, if I'm parsing this correctly.

Reply:
> Can't we parse ECH, determine the zone, and then check to see if ECH is disabled or not (and respond accordingly)?

This seems like it would be much more complicated, and I'm not sure I see the point.

> Also, if ECH is rejected, we need to ignore (and not call checkForResumption()), right? I think that code path is still hit, if I'm parsing this correctly.

Good catch! I added a TODO where this matters. Note that this won't be an issue for our initial deployment since have disabled session tickets.

Reply:
Well, how would we determine if ECH is enabled or disabled without first identifying the zone? Don't we determine the zone by decrypting ECH and looking at the SNI?

Reply:
That seems circular to me. What do you mean "determine if ECH is enabled"?

Reply:
In any case, the true SNI isn't known until we resolve ECH usage. If the server has disabled ECH in its configuration, then ECH will be bypassed and the outer SNI will be used.

Reply:
Right, it's somewhat circular, but that seems unavoidable if we allow ECH to be disabled/enabled on a per-zone basis. (If that's not behavior we currently intend to support, can we just note it and then resolve this comment?)

Reply:
Oh, I think I understand what you mean. The idea is that we should be able reject ECH if we don't recognize the inner SNI. I'm not sure if this is the desired behavior, but I'll leave a note about it.